### PR TITLE
:sparkles: Add support for autoscaling to/from 0

### LINKED ISSUE #1328: Autoscaling to zero
/kind feature

**Describe the solution you'd like**
Auto-scaling to and from zero nodes has been implemented upstream - https://github.com/kubernetes/autoscaler/pull/4840.

We should make sure that we implement the provider side of this contract.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by nikParasyr**
**What this PR does / why we need it**:
Add support for autoscaling to/from 0. To do this:
- Follow the contract for doing this as documented [here](https://cluster-api.sigs.k8s.io/developer/providers/contracts/infra-machine#inframachinetemplate-support-cluster-autoscaling-from-zero)
- Add a new controller to patch OpenStackMachineTemplate.Status accordingly
- Add the feature behind a feature flag

**Which issue(s) this PR fixes**:
Fixes #1328 

**Special notes for your reviewer**:
I'm currently **not** adding support for:
1. Architecture info: Could not find a standardized flavor extra_spec for this one. There are some traits/resource specific ones but they seem to be for when ironic is used. Images can have an `architecture` [attribute](https://docs.openstack.org/glance/latest/admin/useful-image-properties.html#image-property-keys-and-values) but im not sure whether this should be used => I'd prefer to leave this out of this PR and see if users need/require this.
2. GPU info: Again couldn't find the standardized extra_spec for this. I would also need to check what type of info autoscaler expects ( just vendor info or even gpu model info etc )  => I'd prefer to leave this out of this PR and see if users need/require this.

**TODOs**:

- [x] squashed commits
- if necessary:
  - [x] includes documentation
  - [X] adds unit tests
  - [X] adds e2e tests

/hold

### CODE REVIEW

File: pkg/utils/controllers/controllers.go [hunk 1]
```diff
@@ -46,3 +50,15 @@ func ValidateSubnets(subnets []infrav1.Subnet) error {
 	}
 	return nil
 }
+
+func GetInfraCluster(ctx context.Context, c client.Client, cluster *clusterv1.Cluster, namespace string) (*infrav1.OpenStackCluster, error) {
```
Comment (lentzi90) [inline pkg/utils/controllers/controllers.go]: Why not take the namespace from the cluster? In my experience it is safer that way, since there is no possibility to end up with a cross-namespace access then.

**Review by lentzi90 — COMMENTED** [review]
I think we have an RBAC issue. Check this:
```
I0109 15:29:57.597210       1 static_autoscaler.go:602] Scale down status: lastScaleUpTime=2026-01-09 15:18:19.195440928 +0000 UTC m=+10.635138168 lastScaleDownDeleteTime=2026-01-09 15:28:56.89583451 +0000 UTC m=+648.335531740 lastScaleDownFailTime=2026-01-09 14:18:08.994108363 +0000 UTC m=-3599.566194417 scaleDownForbidden=false scaleDownInCooldown=false
I0109 15:29:57.597228       1 static_autoscaler.go:629] Starting scale down
I0109 15:29:57.599020       1 clusterapi_controller.go:770] discovered node group: MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4 (min: 0, max: 2, replicas: 2)
I0109 15:29:57.599030       1 nodes.go:194] autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47 was unneeded for 8m5.499131758s
I0109 15:29:57.599125       1 clusterapi_controller.go:796] node "autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47" is in nodegroup "MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4"
I0109 15:29:57.599136       1 nodes.go:194] autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq was unneeded for 8m5.499131758s
I0109 15:29:57.599214       1 clusterapi_controller.go:796] node "autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq" is in nodegroup "MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4"
I0109 15:29:57.599300       1 clusterapi_controller.go:796] node "autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47" is in nodegroup "MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4"
I0109 15:29:57.599305       1 klogx.go:87] Considering node autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47 for standard scale down
I0109 15:29:57.599385       1 clusterapi_controller.go:796] node "autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq" is in nodegroup "MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4"
I0109 15:29:57.599389       1 klogx.go:87] Considering node autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq for standard scale down
I0109 15:29:57.599469       1 clusterapi_controller.go:796] node "autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47" is in nodegroup "MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4"
I0109 15:29:57.599569       1 clusterapi_controller.go:796] node "autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq" is in nodegroup "MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4"
I0109 15:29:57.599604       1 taints.go:242] Marking node autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47 to be cordoned by Cluster Autoscaler
I0109 15:29:57.599605       1 taints.go:242] Marking node autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq to be cordoned by Cluster Autoscaler
I0109 15:29:57.689862       1 taints.go:223] Successfully added ToBeDeletedByClusterAutoscaler on node autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq
I0109 15:29:57.716675       1 taints.go:223] Successfully added ToBeDeletedByClusterAutoscaler on node autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47
I0109 15:29:57.716701       1 actuator.go:175] Scale-down: removing empty node "autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47"
I0109 15:29:57.716845       1 clusterapi_controller.go:796] node "autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47" is in nodegroup "MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4"
I0109 15:29:57.716876       1 actuator.go:175] Scale-down: removing empty node "autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq"
I0109 15:29:57.717558       1 clusterapi_controller.go:796] node "autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq" is in nodegroup "MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4"
I0109 15:29:57.717797       1 actuator.go:295] Scale-down: waiting 5s before trying to delete nodes
I0109 15:29:57.720009       1 clusterapi_controller.go:770] discovered node group: MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4 (min: 0, max: 2, replicas: 2)
I0109 15:29:57.721597       1 clusterapi_controller.go:770] discovered node group: MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4 (min: 0, max: 2, replicas: 2)
I0109 15:29:57.721773       1 clusterapi_controller.go:796] node "autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47" is in nodegroup "MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4"
I0109 15:29:57.721866       1 clusterapi_controller.go:796] node "autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq" is in nodegroup "MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4"
I0109 15:29:57.723286       1 clusterapi_controller.go:770] discovered node group: MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4 (min: 0, max: 2, replicas: 2)
I0109 15:29:57.793493       1 clusterapi_controller.go:770] discovered node group: MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4 (min: 0, max: 2, replicas: 2)
I0109 15:29:59.187278       1 reflector.go:979] "Watch close" reflector="pkg/mod/k8s.io/client-go@v0.34.2/tools/cache/reflector.go:290" type="*v1.ReplicationController" totalItems=7
I0109 15:30:00.093318       1 reflector.go:979] "Watch close" reflector="pkg/mod/k8s.io/client-go@v0.34.2/tools/cache/reflector.go:290" type="*v1.Namespace" totalItems=6
I0109 15:30:00.231206       1 reflector.go:979] "Watch close" reflector="pkg/mod/k8s.io/client-go@v0.34.2/tools/cache/reflector.go:290" type="*v1.CSIStorageCapacity" totalItems=8
I0109 15:30:02.718026       1 drain.go:140] All pods removed from autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq
I0109 15:30:02.718091       1 drain.go:140] All pods removed from autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47
I0109 15:30:02.718215       1 clusterapi_controller.go:796] node "autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47" is in nodegroup "MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4"
I0109 15:30:02.718219       1 clusterapi_controller.go:796] node "autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq" is in nodegroup "MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4"
I0109 15:30:02.720425       1 clusterapi_controller.go:796] node "autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47" is in nodegroup "MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4"
I0109 15:30:02.720552       1 clusterapi_controller.go:796] node "autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47" is in nodegroup "MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4"
E0109 15:30:02.722830       1 delete_in_batch.go:206] Scale-down: couldn't delete empty node "autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47", , status error: failed to delete nodes from group MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4: machines.cluster.x-k8s.io "autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47" is forbidden: User "system:serviceaccount:autoscaler-fqibb8:cluster-autoscaler" cannot update resource "machines" in API group "cluster.x-k8s.io" in the namespace "autoscaler-fqibb8"
I0109 15:30:02.724397       1 clusterapi_controller.go:796] node "autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq" is in nodegroup "MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4"
I0109 15:30:02.724510       1 clusterapi_controller.go:796] node "autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq" is in nodegroup "MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4"
E0109 15:30:02.726369       1 delete_in_batch.go:206] Scale-down: couldn't delete empty node "autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq", , status error: failed to delete nodes from group MachineDeployment/autoscaler-fqibb8/autoscaler-ywetms-md-0-jbkw4: machines.cluster.x-k8s.io "autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq" is forbidden: User "system:serviceaccount:autoscaler-fqibb8:cluster-autoscaler" cannot update resource "machines" in API group "cluster.x-k8s.io" in the namespace "autoscaler-fqibb8"
I0109 15:30:02.753589       1 taints.go:323] Releasing taint {Key:ToBeDeletedByClusterAutoscaler Value:1767972597 Effect:NoSchedule TimeAdded:<nil>} on node autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47
I0109 15:30:02.753612       1 taints.go:343] Marking node autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47 to be uncordoned by Cluster Autoscaler
I0109 15:30:02.756916       1 taints.go:323] Releasing taint {Key:ToBeDeletedByClusterAutoscaler Value:1767972597 Effect:NoSchedule TimeAdded:<nil>} on node autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq
I0109 15:30:02.756932       1 taints.go:343] Marking node autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq to be uncordoned by Cluster Autoscaler
I0109 15:30:02.816827       1 taints.go:358] Successfully released ToBeDeletedByClusterAutoscaler on node autoscaler-ywetms-md-0-jbkw4-fchnc-nxl47
I0109 15:30:02.847341       1 taints.go:358] Successfully released ToBeDeletedByClusterAutoscaler on node autoscaler-ywetms-md-0-jbkw4-fchnc-b8flq
```

**Comment by lentzi90** [comment]
So that is the RBAC of the autoscaler itself. Do we need to add some Role(Binding) for this use-case?

**Comment by nikParasyr** [comment]
> So that is the RBAC of the autoscaler itself. Do we need to add some Role(Binding) for this use-case?

Not on capo. The autoscaler role(binding) is set on the e2e tests. I’ll check tomorrow what I missed

**Comment by lentzi90** [comment]
Great! It is passing! Let's check also the full suite for any edge cases.
/test pull-cluster-api-provider-openstack-e2e-full-test

File: pkg/utils/controllers/controllers.go (hunk 1 above)
Comment (nikParasyr) [reply pkg/utils/controllers/controllers.go]: fixed

**Comment by nikParasyr** [comment]
@lentzi90 converting this to draft as autoscaling can also work on clusters without ClusterClass. ill start implementing it now ( shouldnt be too much work )

**Comment by nikParasyr** [comment]
@lentzi90 this now includes non CC clusters with the latest commit. Also updated docs etc.
This is _finally_ ready from my side

**Comment by nikParasyr** [comment]
> Great! This looks good to me. Could you please squash the commits?

Done

**Review by lentzi90 — COMMENTED** [review]
/approve
I think we may want to move the autoscaler test to the optional "full" suite going forward but I am happy with it like this for now. It seems to add ~10 minutes so not too bad.

**Comment by lentzi90** [comment]
Let's take it in. I'll start on the RC
/hold cancel
